graph/src/components/ethereum/network.rs

```diff
@@ -0,0 +1,273 @@
+use failure::{format_err, Error};
+use std::cmp::Ord;
+use std::collections::{BTreeMap, HashMap};
+use std::fmt;
+use std::str::FromStr;
+use std::sync::Arc;
+
+use crate::components::ethereum::EthereumAdapter;
+pub use crate::impl_slog_value;
+
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub enum NetworkCapability {
+    Full,
+    Archive,
+    Traces,
+}
+
+impl FromStr for NetworkCapability {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(match s {
+            "full" => NetworkCapability::Full,
+            "archive" => NetworkCapability::Archive,
+            "traces" => NetworkCapability::Traces,
+            s => return Err(format_err!("invalid network capability provided: {}", s)),
+        })
+    }
+}
+
+impl fmt::Display for NetworkCapability {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            NetworkCapability::Full => write!(f, "full"),
+            NetworkCapability::Archive => write!(f, "archive"),
+            NetworkCapability::Traces => write!(f, "traces"),
+        }
+    }
+}
+
+impl NetworkCapability {
+    pub fn sufficient_capability(
+        network: &Vec<NetworkCapability>,
+        required: &Vec<NetworkCapability>,
+    ) -> bool {
+        match required[..] {
+            [NetworkCapability::Full] => true,
+            [NetworkCapability::Traces] => network.contains(&NetworkCapability::Traces),
+            [NetworkCapability::Full, NetworkCapability::Traces] => {
+                network.contains(&NetworkCapability::Traces)
+            }
+            [NetworkCapability::Archive] => network.contains(&NetworkCapability::Archive),
+            [NetworkCapability::Archive, NetworkCapability::Traces] => {
+                network == &vec![NetworkCapability::Archive, NetworkCapability::Traces]
+            }
+            _ => false,
+        }
+    }
+
+    pub fn cheapest() -> Self {
+        NetworkCapability::Full
+    }
+}
+
+impl_slog_value!(NetworkCapability, "{:?}");
+
+#[derive(Clone)]
+pub struct EthereumNetworkAdapters {
+    pub adapters: BTreeMap<Vec<NetworkCapability>, Arc<dyn EthereumAdapter>>,
```

Comment:
I think this is too complicated. We're going to deal with a handful of adapters perhaps. A simple
```rust
struct EthereumNetworkAdapter {
  capabilities: Vec<NetworkCapability>,
  adapter: dyn EthereumAdapter
}
```
and
```rust
  pub adapters: Vec<EthereumNetworkAdapter>,
```
should be sufficient.

Reply:
I was hoping to move toward that, but right now I wanted to use the BTreeMap to enforce uniqueness in (network_name, capabilities) since there currently isn't a selection algorithm for deciding between identical networks.  

Ok, I see below that you suggest randomly picking if there are two, this does open up our options to supply more networks to the node.  Works for me.   It could also be useful to use identical network capabilities as fallback networks, restarting the subgraph with another endpoint if it gets stuck in a retry loop past a threshold.

Reply:
Yeah, I don't think there are any uniqueness constraints here.